### DISCUSSION TITLE
Polling pattern

### DISCUSSION THREAD

**Discussion by sakulstra** [pmndrs/zustand#1102] [Q&A]
Hey i've just started looking into `zustand` and am wondering if there's any good pattern for async polling as i don't see a simple way of keeping track of current listeners of a certain value.

Let's imagine I have  a simple store like this, where I need to poll as long and only as long there's a subscriber to `data`
```js
create(set => ({
  data: {}
  poll: () => doSthAsync().then(data => set({data})),
}))
```

I guess i could create a `usePoll()` hook which periodically polls the data (and stops once it's unmounted), but I imagine that would get tricky as soon as there might be different consumers that potentially unnecessarily poll the same source.
How are people solving this issue?

**Comment by dai-shi** [comment]
If this is global state, you might not need to do it in react lifecycle.
In that case you can do it outside.
```js
const useDataStore = create(set => ({
  data: {}
  poll: () => doSthAsync().then(data => set({data})),
}))

const id = setTimeout(() => useDataStore().getState().poll(), 5000)
// use clearTimeout if necessary
```

**Comment by sakulstra** [comment]
@dai-shi, but when it is?

I guess in essence I want to recreate sth like react-query with Zustand where multiple consumers subscribe to the same data and "share" the polling/cache, but when all components are unmounted the polling should stop as well.

**Reply by dai-shi** [reply]
So, you want to use it with react lifecycle. You might need to track the number of mounted components.

```js
const useDataStore = create((set, get) => ({
  data: {},
  mounted: 0,
  timer: null,
  mount: () => {
    const { mounte, timer } = get()
    set({ mounted: mounted + 1 })
    if (mounted) {
      const timer = setTimeout(() => doSthAsync().then(data => set({data})), 1000)
      set({ timer })
    }
  },
  unmount: () => {
    const { mounte, timer } = get()
    set({ mounted: mounted - 1 })
    if (mounted === 1) {
      clearTimeout(timer)
      set({ timer: null })
    }
  },
}))

**Reply by sakulstra** [reply]
hmm, yah sth like this should work thx :+1:

Still wondering if there could be a nicer more generic way via some `onSubscribe`  and `onDelete` hook, but i guess that kinda contradicts the [simplicity of zustand](https://github.com/pmndrs/zustand/blob/main/src/vanilla.ts#L136)
I guess i could use a middleware similar to https://github.com/pmndrs/zustand/blob/main/src/middleware/subscribeWithSelector.ts#L59 to alter the subscribe and delete behavior :thinking:

**Reply by dai-shi** [reply]
Middleware should work in such cases. I can recommend it for JS users, but typing with TS would be extremely hard.

I prefer hook composition in that case.
```js
const useDataStoreInternal = create(set => ({
  data: {}
}))

export const useDataStore = (selector) => {
  const ref = useRef({ mounted: 0, timer: null })
  useEffect(() => {
    if (ref.current.mounted === 0) {
      ref.current.timer = setTimeout(() => {
        doSthAsync().then(data => useDataStoreInternal.setState({data}))
      }, 1000)
    }
    ++ref.current.mounted
    return () => {
      --ref.current.mounted
      if (ref.current.mounted === 0) {
        clearTimeout(ref.current.timer)
      }
    }      
    }
  }, [])
  return useDataStoreInternal(selector)
}
```
